proto/message.go

```diff
@@ -0,0 +1,261 @@
+	return f.Uint64(uint64(v))
+}
+
+func (f FieldNumber) Uint64(v uint64) RawMessage {
+	return AppendVarint(nil, f, v)
+}
+
+func (f FieldNumber) Fixed32(v uint32) RawMessage {
+	return AppendFixed32(nil, f, v)
+}
+
+func (f FieldNumber) Fixed64(v uint64) RawMessage {
+	return AppendFixed64(nil, f, v)
+}
+
+func (f FieldNumber) Float32(v float32) RawMessage {
+	return f.Fixed32(math.Float32bits(v))
+}
+
+func (f FieldNumber) Float64(v float64) RawMessage {
+	return f.Fixed64(math.Float64bits(v))
+}
+
+func (f FieldNumber) String(v string) RawMessage {
+	return f.Bytes([]byte(v))
+}
+
+func (f FieldNumber) Bytes(v []byte) RawMessage {
+	return AppendVarlen(nil, f, v)
+}
+
+// Value constructs a RawMessage for field number f from v.
+func (f FieldNumber) Value(v interface{}) RawMessage {
+	switch x := v.(type) {
+	case bool:
+		return f.Bool(x)
+	case int:
+		return f.Int(x)
+	case int32:
+		return f.Int32(x)
+	case int64:
+		return f.Int64(x)
+	case uint:
+		return f.Uint(x)
+	case uint32:
+		return f.Uint32(x)
+	case uint64:
+		return f.Uint64(x)
+	case float32:
+		return f.Float32(x)
+	case float64:
+		return f.Float64(x)
+	case string:
+		return f.String(x)
+	case []byte:
+		return f.Bytes(x)
+	default:
+		panic("cannot rewrite value of unsupported type")
+	}
+}
+
+// The WireType enumeration represents the different protobuf wire types.
+type WireType int
+
+const (
+	Varint  WireType = 0
+	Fixed64 WireType = 1
+	Varlen  WireType = 2
+	Fixed32 WireType = 5
+)
+
+func (wt WireType) String() string {
+	return wireType(wt).String()
+}
+
+func Append(m RawMessage, f FieldNumber, t WireType, v []byte) RawMessage {
+	b := [24]byte{}
+	n, _ := encodeVarint(b[:], EncodeTag(f, t))
+	if t == Varlen {
+		n1, _ := encodeVarint(b[n:], uint64(len(v)))
+		n += n1
+	}
+	m = append(m, b[:n]...)
+	m = append(m, v...)
+	return m
+}
+
+func AppendVarint(m RawMessage, f FieldNumber, v uint64) RawMessage {
+	b := [12]byte{}
+	n, _ := encodeVarint(b[:], v)
+	return Append(m, f, Varint, b[:n])
+}
+
+func AppendVarlen(m RawMessage, f FieldNumber, v []byte) RawMessage {
+	return Append(m, f, Varlen, v)
+}
+
+func AppendFixed32(m RawMessage, f FieldNumber, v uint32) RawMessage {
+	b := [4]byte{}
+	binary.LittleEndian.PutUint32(b[:], v)
+	return Append(m, f, Fixed32, b[:])
+}
+
+func AppendFixed64(m RawMessage, f FieldNumber, v uint64) RawMessage {
+	b := [8]byte{}
+	binary.LittleEndian.PutUint64(b[:], v)
+	return Append(m, f, Fixed64, b[:])
+}
+
+func Parse(m RawMessage) (FieldNumber, WireType, RawValue, RawMessage, error) {
+	tag, n, err := decodeVarint(m)
+	if err != nil {
+		return 0, 0, nil, m, fmt.Errorf("decoding protobuf field number: %w", err)
+	}
+	m = m[n:]
+	f, t := DecodeTag(tag)
+
+	switch t {
+	case Varint:
+		_, n, err := decodeVarint(m)
+		if err != nil {
+			return f, t, nil, m, fmt.Errorf("decoding varint field %d: %w", f, err)
+		}
+		return f, t, RawValue(m[:n]), m[n:], nil
+
+	case Varlen:
+		l, n, err := decodeVarint(m) // length
```

Comment:
Could we add another check here that `l < uint64(len(m) - n)`? so that we know `n+int(l)` won't panic, or give an incorrect result if `int(l)` truncates.